### PR TITLE
fix(security): isolate PR summary publish evidence

### LINKED ISSUE #3412: Isolate PR summary rendering from write-scoped GitHub token permissions
# Isolate PR summary rendering from write-scoped GitHub token permissions

## Summary

`pr-ci-status-comment.yml` renders and publishes the PR summary in the same pull-request job context that grants issue and pull-request write permissions. For non-fork PRs, the job checks out repository content and runs the checked-out PR summary renderer before using the GitHub API to create or update the bot comment.

## Evidence

- `repo/.github/workflows/pr-ci-status-comment.yml:3-8` defines pull_request, schedule, and workflow_dispatch entry points.
- `repo/.github/workflows/pr-ci-status-comment.yml:33-37` grants `issues: write` and `pull-requests: write` at workflow scope.
- `repo/.github/workflows/pr-ci-status-comment.yml:206-211` runs the `summarize` job for non-fork pull requests.
- `repo/.github/workflows/pr-ci-status-comment.yml:230` checks out repository content, and `:280-286` runs `node scripts/summary/render-pr-summary.mjs` from that workspace.
- `repo/.github/workflows/pr-ci-status-comment.yml:477-506` appends generated Markdown artifacts to `artifacts/summary/PR_SUMMARY.md`.
- `repo/.github/workflows/pr-ci-status-comment.yml:556-570` publishes that file through `github.rest.issues.updateComment` or `github.rest.issues.createComment`.

No workflow was run, no token value was read, and no GitHub API mutation was performed during the audit.

## Risk

A same-repository PR author, compromised contributor account, or configuration that allows untrusted non-fork branches can influence code and Markdown handled by a write-permission PR maintenance job. Public fork reachability is not asserted because the workflow excludes fork heads. Practical write impact depends on live GitHub Actions token behavior and repository protections.

## Recommended remediation

- Split PR-controlled rendering from comment publishing.
- Run the rendering job with read-only permissions and `actions/checkout` credential persistence disabled.
- Publish comments from a separate trusted-ref job that consumes a sanitized artifact and has only the minimum issue-comment permission required.
- Avoid running checked-out local actions, package scripts, or mutable renderer code in jobs that have issue, pull-request, checks, security-events, or contents write permissions.
- Add workflow linting for pull_request jobs that combine write permissions with local script execution.

### LINKED ISSUE #3414: Bind auto-merge Change Package status to trusted validation artifacts
# Bind auto-merge Change Package status to trusted validation artifacts

## Summary

The auto-merge enabler currently treats Change Package Validation status in an AE PR summary comment as trusted when the comment author is `github-actions[bot]`. The status is parsed from Markdown, not from a structured artifact or a workflow run bound to the PR head SHA.

## Security impact

A same-repository PR workflow path can generate and publish PR summary Markdown from checked-out PR content. Auto-merge later consumes that comment as merge-governance evidence. If attacker-influenced summary content shadows or duplicates the validation section, the change-package gate can be treated as passed or warning-only while the actual generated validation content would not support that result.

This does not assert that every PR can be merged automatically. Auto-merge still depends on repository variables, labels, required checks, review policy, token scope, and branch/ruleset settings. The issue is that one of those gates is authenticated by bot comment author and Markdown shape rather than trusted validation provenance.

## Evidence

- `repo/.github/workflows/pr-ci-status-comment.yml:206-211` runs PR summary generation for non-fork pull requests.
- `repo/.github/workflows/pr-ci-status-comment.yml:230-286` checks out the workspace and runs local summary rendering code.
- `repo/.github/workflows/pr-ci-status-comment.yml:556-570` posts the resulting summary as an `<!-- AE-PR-SUMMARY -->` comment.
- `repo/scripts/ci/lib/change-package-gate.mjs:1-64` trusts `github-actions` / `github-actions[bot]` comments and parses the Markdown validation section from the comment body.
- `repo/scripts/ci/auto-merge-enabler.mjs:332-360` consumes that parsed status, and `:370-374` can proceed to auto-merge enablement when no other blockers remain.

No GitHub API call, workflow execution, token access, live exploitation, or repository source modification was performed during validation.

## Suggested remediation

Use a structured Change Package Validation artifact or check run produced by trusted base-ref code and bind it to the PR head SHA and workflow run. Have auto-merge consume that structured source rather than PR summary Markdown. If comments remain an input, reject duplicate validation sections, use a unique machine-readable generated block, verify comment/run provenance, and fail closed on ambiguity.

## Regression test idea

Add tests for duplicate validation sections, stale bot comments, untrusted authors, and head-SHA/run mismatches. The auto-merge gate should fail closed unless the validation evidence is unambiguous and provenance-bound.

### PR DESCRIPTION
**Pull request by ootakazuhiko**
## Summary

Closes #3412.
Closes #3414.

- Splits PR-controlled summary rendering from write-scoped PR comment publication.
- Keeps the `summarize` job read-only, removes write-scoped secrets from local summary/harness script execution, and publishes comments from a checkout-free `publish-summary` job.
- Publishes a head-SHA-bound `Change Package Validation` check run from the structured validation artifact.
- Changes auto-merge gating to consume the PR-head check-run evidence instead of parsing Change Package Validation status from PR summary Markdown.
- Adds regression tests for workflow permission boundaries and fail-closed Change Package check-run interpretation.

## Verification

- `pnpm install --offline --frozen-lockfile`
- `pnpm -s exec vitest run tests/unit/ci/change-package-gate.test.ts tests/unit/ci/workflow-permission-boundary.test.ts --reporter dot`
- `/home/devuser/work/CodeX/ae-frameworkA/.codex-local/bin/actionlint-bin .github/workflows/pr-ci-status-comment.yml`
- `pnpm -s exec eslint --quiet scripts/ci/lib/change-package-gate.mjs scripts/ci/auto-merge-enabler.mjs tests/unit/ci/change-package-gate.test.ts tests/unit/ci/workflow-permission-boundary.test.ts`
- `pnpm -s run types:check`
- `pnpm -s run build`
- `node scripts/ci/validate-json.mjs`
- `pnpm -s run check:schemas`
- `pnpm -s run check:doc-consistency`
- `git diff --check`

## Acceptance

- PR summary rendering can execute checked-out PR code only with read-scoped permissions.
- PR summary/progress comments are published by a separate job that does not checkout or execute repository-local scripts.
- Auto-merge no longer trusts Change Package status parsed from `<!-- AE-PR-SUMMARY -->` Markdown.
- Auto-merge fails closed for missing, pending, failed, or ambiguous `Change Package Validation` check-run evidence.

## Rollback

Revert this PR to restore the previous single-job summary comment publication and Markdown-based auto-merge Change Package gate. That rollback would re-open #3412 and #3414.

### CODE REVIEW

File: .github/workflows/pr-ci-status-comment.yml [hunk 1]
```diff
@@ -737,8 +850,14 @@ jobs:
       && (inputs.mode == 'eligibility')
       && (inputs.pr_number != '')
     runs-on: ubuntu-latest
+    permissions:
+      contents: read
+      issues: write
+      pull-requests: write
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply .github/workflows/pr-ci-status-comment.yml:1074]: Addressed in `544d2094`: the workflow-dispatch eligibility path now uses `resolveChangePackageValidationStatusFromChecks(pr.statusCheckRollup || [])`, removed PR-summary comment parsing from `auto-merge-eligible.mjs`, and fails closed for missing, pending, ambiguous, or failed Change Package Validation check-run evidence.

File: scripts/ci/lib/change-package-gate.mjs [hunk 2]
```diff
@@ -44,6 +48,82 @@ function parseChangePackageValidationResult(body) {
...
+  return Number.isFinite(ts) ? ts : 0;
+}
+
+function normalizeCheckRunName(checkRun) {
+  return String(checkRun?.name ?? '').trim();
+}
+
+function isChangePackageValidationCheck(checkRun) {
+  return CHANGE_PACKAGE_VALIDATION_CHECK_NAMES.has(normalizeCheckRunName(checkRun));
+}
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply scripts/ci/lib/change-package-gate.mjs:70]: Addressed in `544d2094`: Change Package Validation check-run names are normalized to lower case before allowlist matching, preserving the allowed-name set while making the comparison case-insensitive. Added a mixed-case regression assertion.

File: .github/workflows/pr-ci-status-comment.yml [hunk 3]
```diff
@@ -643,6 +629,122 @@ jobs:
...
+            const upsertComment = async (marker, body) => {
+              const payload = limitComment(`${marker}\n\n${body}`, marker);
+              const comments = await github.rest.issues.listComments({ owner, repo, issue_number: number, per_page: 100 });
+              const existing = comments.data.find((comment) => typeof comment.body === 'string' && comment.body.includes(marker));
+              if (existing) {
+                await github.rest.issues.updateComment({ owner, repo, comment_id: existing.id, body: payload });
+              } else {
+                await github.rest.issues.createComment({ owner, repo, issue_number: number, body: payload });
+              }
+            };
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply .github/workflows/pr-ci-status-comment.yml:882]: Addressed in `544d2094`: the checkout-free publisher now paginates all PR issue comments before selecting the existing AE marker comment, so marker updates are not limited to the first 100 comments.

File: .github/workflows/pr-ci-status-comment.yml [hunk 4]
```diff
@@ -643,6 +629,122 @@ jobs:
...
+            const progressJsonText = readText('summary.json');
+            if (progressBody && progressJsonText) {
+              const progressJson = JSON.parse(progressJsonText);
+              await upsertComment(
+                '<!-- AE-PR-PROGRESS -->',
+                `${progressBody}\n\n<!-- AE-PR-PROGRESS-JSON ${JSON.stringify(progressJson)} -->`,
+              );
+            } else {
+              core.warning('Progress summary files were not both present in the publish artifact.');
+            }
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply .github/workflows/pr-ci-status-comment.yml]: Addressed in `544d2094`: the progress JSON block is included only when the full marker/comment payload fits the publication cap. If the JSON block would exceed the cap, the publisher emits the human-readable progress Markdown without embedding invalid truncated JSON and logs a warning.

File: .github/workflows/pr-ci-status-comment.yml [hunk 5]
```diff
@@ -643,6 +629,149 @@ jobs:
...
+            const upsertComment = async (marker, body) => {
+              const payload = limitComment(`${marker}\n\n${body}`, marker);
+              const comments = await listAllIssueComments();
+              const existing = comments.find((comment) => typeof comment.body === 'string' && comment.body.includes(marker));
+              if (existing) {
+                await github.rest.issues.updateComment({ owner, repo, comment_id: existing.id, body: payload });
+              } else {
+                await github.rest.issues.createComment({ owner, repo, issue_number: number, body: payload });
+              }
+            };
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply .github/workflows/pr-ci-status-comment.yml:882]: Addressed in `b5ef9515`: `upsertComment` now only updates existing comments authored by `github-actions` / `github-actions[bot]` whose body starts with the requested AE marker, avoiding updates to arbitrary user comments that merely quote or mention a marker.

File: .github/workflows/pr-ci-status-comment.yml [hunk 6]
```diff
@@ -643,6 +629,149 @@ jobs:
...
+              if (progressCommentBody === progressBody) {
+                core.warning('Progress summary JSON omitted because it would exceed the PR comment size cap.');
+              }
+              await upsertComment(
+                '<!-- AE-PR-PROGRESS -->',
+                progressCommentBody,
+              );
+            } else {
+              core.warning('Progress summary files were not both present in the publish artifact.');
+            }
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply .github/workflows/pr-ci-status-comment.yml:920]: Addressed in `b5ef9515`: malformed progress `summary.json` is now non-fatal. The publisher logs a warning and still publishes the human-readable progress Markdown without the embedded JSON block, so Change Package check-run publication can continue.

File: .github/workflows/pr-ci-status-comment.yml [hunk 7]
```diff
@@ -643,6 +629,168 @@ jobs:
...
+              const filePath = path.join(artifactRoot, relativePath);
+              if (!fs.existsSync(filePath)) return null;
+              return fs.readFileSync(filePath, 'utf8').replace(/\u0000/g, '\uFFFD');
+            };
+            const limitComment = (text, label) => {
+              const bytes = Buffer.byteLength(text, 'utf8');
+              if (bytes <= MAX_COMMENT_BYTES) return text;
+              const truncated = Buffer.from(text, 'utf8').subarray(0, MAX_COMMENT_BYTES).toString('utf8');
+              return `${truncated}\n\n_[${label} truncated to ${MAX_COMMENT_BYTES} bytes before publication.]_`;
+            };
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply .github/workflows/pr-ci-status-comment.yml:833]: Addressed in `053e0014`: `limitComment` now reserves byte budget for the truncation notice before truncating the UTF-8 payload, so the final published comment remains within `MAX_COMMENT_BYTES` including the suffix.

File: .github/workflows/pr-ci-status-comment.yml [hunk 8]
```diff
@@ -643,6 +629,168 @@ jobs:
...
+              );
+            } else {
+              core.warning('Progress summary Markdown was not present in the publish artifact.');
+            }
+
+            const readJson = (relativePath) => {
+              const text = readText(relativePath);
+              if (!text) return null;
+              return JSON.parse(text);
+            };
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply .github/workflows/pr-ci-status-comment.yml]: Addressed in `053e0014`: `readJson` now catches parse failures, logs a warning, and treats malformed structured artifacts as missing, allowing summary/progress publication and the Change Package check-run creation to continue fail-closed.

File: scripts/ci/lib/change-package-gate.mjs [hunk 9]
```diff
@@ -44,6 +48,82 @@ function parseChangePackageValidationResult(body) {
...
+      return 'warn';
+    case 'FAILURE':
+    case 'CANCELLED':
+    case 'TIMED_OUT':
+    case 'ACTION_REQUIRED':
+    case 'STALE':
+      return 'fail';
+    default:
+      return 'missing';
+  }
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply scripts/ci/lib/change-package-gate.mjs:103]: Addressed in `053e0014`: `SKIPPED` is now handled explicitly as fail-closed Change Package Validation check-run evidence, with a regression assertion covering the GitHub `SKIPPED` conclusion.

File: .github/workflows/pr-ci-status-comment.yml [hunk 10]
```diff
@@ -204,6 +202,9 @@ jobs:
       && github.event_name == 'pull_request'
       && !github.event.pull_request.head.repo.fork
     runs-on: ubuntu-latest
+    permissions:
+      issues: write
+      pull-requests: read
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply .github/workflows/pr-ci-status-comment.yml]: Addressed in `053e0014`: the `label` job permissions now include only `issues: write`; the unnecessary `pull-requests: read` scope was removed and the workflow permission-boundary test asserts the narrower scope.

File: .github/workflows/pr-ci-status-comment.yml [hunk 11]
```diff
@@ -643,6 +628,186 @@ jobs:
...
+            const upsertComment = async (marker, body) => {
+              const payload = limitComment(`${marker}\n\n${body}`, marker);
+              const comments = await listAllIssueComments();
+              const existing = comments.find((comment) => isTrustedAutomationComment(comment, marker));
+              if (existing) {
+                await github.rest.issues.updateComment({ owner, repo, comment_id: existing.id, body: payload });
+              } else {
+                await github.rest.issues.createComment({ owner, repo, issue_number: number, body: payload });
+              }
+            };
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply .github/workflows/pr-ci-status-comment.yml:882]: Addressed in `3f5816d8`: the publisher now caches the paginated issue-comment list once per run through `getIssueComments()`, and `upsertComment` selects the most recent trusted automation marker comment via reverse search before updating it. The workflow permission-boundary test now asserts both behaviors.

File: .github/workflows/pr-ci-status-comment.yml [hunk 12]
```diff
@@ -643,6 +628,193 @@ jobs:
...
+              const login = String(comment?.user?.login || '').trim().toLowerCase();
+              const body = typeof comment?.body === 'string' ? comment.body : '';
+              return (
+                (login === 'github-actions' || login === 'github-actions[bot]')
+                && body.startsWith(marker)
+              );
+            };
+            const upsertComment = async (marker, body) => {
+              const payload = limitComment(`${marker}\n\n${body}`, marker);
+              const comments = await getIssueComments();
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply .github/workflows/pr-ci-status-comment.yml:875]: Addressed in `7760da72`: `upsertComment` now derives a display label from the HTML marker before calling `limitComment`, so truncation notices use visible labels such as `AE-PR-SUMMARY` instead of the raw `<!-- ... -->` marker.

File: .github/workflows/pr-ci-status-comment.yml [hunk 13]
```diff
@@ -643,6 +628,193 @@ jobs:
...
+              repo,
+              name: 'Change Package Validation',
+              head_sha: headSha,
+              status: 'completed',
+              conclusion,
+              output: {
+                title,
+                summary,
+              },
+            });
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply .github/workflows/pr-ci-status-comment.yml]: Addressed in `7760da72`: the workflow now adds a read-only `validate-change-package` job that checks out `github.event.pull_request.base.sha`, downloads the PR-produced change package artifact, re-runs `scripts/change-package/validate.mjs` from trusted base-ref code, and uploads a trusted validation artifact. The checkout-free writer job now downloads that trusted artifact and uses it as the sole source for the `Change Package Validation` check-run conclusion.

File: .github/workflows/pr-ci-status-comment.yml [hunk 14]
```diff
@@ -643,6 +630,337 @@ jobs:
...
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          ref: ${{ github.event.pull_request.base.sha }}
+          persist-credentials: false
+      - name: Setup Node + pnpm
+        uses: ./.github/actions/setup-node-pnpm
+        with:
+          node-version: '20'
+      - name: Install trusted validation dependencies
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply .github/workflows/pr-ci-status-comment.yml:656]: Addressed in `1a7ca7c9` and retained in latest head `45fea622`: `validate-change-package` now runs the shared Node/pnpm setup and installs trusted base-ref dependencies with `pnpm install --frozen-lockfile --config.use-lockfile=true --config.package-lock=true` before invoking `scripts/change-package/validate.mjs`. The workflow boundary test asserts the install step, and the latest CI `validate-change-package` job passes on the PR head.

File: .github/workflows/pr-ci-status-comment.yml [hunk 15]
```diff
@@ -235,6 +235,11 @@ jobs:
       && !github.event.pull_request.head.repo.fork
       && github.event.action != 'edited'
     runs-on: ubuntu-latest
+    permissions:
+      actions: read
+      contents: read
+      issues: read
+      pull-requests: read
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply .github/workflows/pr-ci-status-comment.yml:242]: Addressed in `45fea622`: the read-only `summarize` checkout now sets `persist-credentials: false`, reducing accidental token exposure to PR-controlled code. Added a workflow permission-boundary regression assertion for the `summarize` checkout configuration.

File: scripts/ci/lib/change-package-gate.mjs [hunk 16]
```diff
@@ -44,6 +48,83 @@ function parseChangePackageValidationResult(body) {
...
+      return 'pass';
+    case 'NEUTRAL':
+      return 'warn';
+    case 'FAILURE':
+    case 'CANCELLED':
+    case 'SKIPPED':
+    case 'TIMED_OUT':
+    case 'ACTION_REQUIRED':
+    case 'STALE':
+      return 'fail';
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply scripts/ci/lib/change-package-gate.mjs:100]: Addressed in `45fea622`: `STARTUP_FAILURE` is now explicitly mapped to fail-closed Change Package Validation evidence instead of falling through to `missing`. Added a regression assertion covering a completed `STARTUP_FAILURE` check run.

File: scripts/ci/lib/change-package-gate.mjs [hunk 17]
```diff
@@ -44,6 +48,84 @@ function parseChangePackageValidationResult(body) {
...
+      return 'missing';
+  }
+}
+
+function resolveChangePackageValidationStatusFromChecks(checkRuns = []) {
+  if (!Array.isArray(checkRuns) || checkRuns.length === 0) {
+    return { status: 'missing', sourceUrl: null };
+  }
+
+  const candidates = checkRuns.filter(isChangePackageValidationCheck);
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply scripts/ci/lib/change-package-gate.mjs:111]: Addressed in latest head `f80b3ada`: Change Package Validation candidates are now restricted to `__typename === 'CheckRun'` before name matching, so similarly named `StatusContext` rollup entries are ignored. Added a regression assertion for a same-name `StatusContext` returning `missing` rather than being consumed as check-run evidence.

File: scripts/ci/lib/change-package-gate.mjs [hunk 18]
```diff
@@ -44,6 +48,84 @@ function parseChangePackageValidationResult(body) {
...
+    case 'CANCELLED':
+    case 'SKIPPED':
+    case 'TIMED_OUT':
+    case 'STARTUP_FAILURE':
+    case 'ACTION_REQUIRED':
+    case 'STALE':
+      return 'fail';
+    default:
+      return 'missing';
+  }
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply scripts/ci/lib/change-package-gate.mjs:103]: Addressed in latest head `f80b3ada`: a completed Change Package Validation check run with an unknown conclusion now maps to fail-closed `fail` instead of `missing`, reflecting that evidence exists but is not interpretable as pass/warn. Added a regression assertion for an unknown conclusion.

File: scripts/ci/lib/change-package-gate.mjs [hunk 19]
```diff
@@ -44,6 +48,87 @@ function parseChangePackageValidationResult(body) {
...
+  if (latestStatuses.size > 1) {
+    return { status: 'ambiguous', sourceUrl: null };
+  }
+
+  const latest = latestCandidates[latestCandidates.length - 1];
+  return {
+    status: mapCheckRunToChangePackageStatus(latest),
+    sourceUrl: checkRunSourceUrl(latest),
+  };
+}
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply scripts/ci/lib/change-package-gate.mjs:139]: Addressed in latest head `5444b33d`: `resolveChangePackageValidationStatusFromChecks` now explicitly treats timestamp-less pending Change Package Validation check runs as `pending` before considering older timestamped completed evidence. This keeps auto-merge fail-closed while a newer queued/in-progress validation rerun is present.

File: tests/unit/ci/change-package-gate.test.ts [hunk 20]
```diff
@@ -77,4 +78,127 @@ describe('change-package gate helpers', () => {
...
+    });
+    expect(resolveChangePackageValidationStatusFromChecks([
+      {
+        __typename: 'CheckRun',
+        name: 'Change Package Validation',
+        status: 'IN_PROGRESS',
+        conclusion: '',
+        startedAt: '2026-06-04T00:00:00Z',
+      },
+    ])).toEqual({ status: 'pending', sourceUrl: null });
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply tests/unit/ci/change-package-gate.test.ts:172]: Addressed in latest head `5444b33d`: added a regression assertion where an older completed successful Change Package Validation check is followed by a timestamp-less `QUEUED` rerun; the helper now returns `pending` rather than selecting the older pass.

File: scripts/ci/lib/change-package-gate.mjs [hunk 21]
```diff
@@ -44,6 +48,96 @@ function parseChangePackageValidationResult(body) {
...
+
+function checkRunSourceUrl(checkRun) {
+  return typeof checkRun?.detailsUrl === 'string'
+    ? checkRun.detailsUrl
+    : (typeof checkRun?.details_url === 'string'
+      ? checkRun.details_url
+      : (typeof checkRun?.html_url === 'string'
+        ? checkRun.html_url
+        : (typeof checkRun?.url === 'string' ? checkRun.url : null)));
+}
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply scripts/ci/lib/change-package-gate.mjs:80]: Addressed in latest head `9109a850`: `checkRunSourceUrl` now scans the ordered URL field list (`detailsUrl`, `details_url`, `html_url`, `url`) in a simple loop and returns the first non-empty string, preserving field priority while removing the nested ternary.
